### ISSUE TITLE
Calling EventSource.Write with data containing an array of string throws on .net Core 2.1, works on .net Core 2.0

### ISSUE THREAD

**Issue by smmalis37** [dotnet/coreclr#18225]
I have code that calls EventSource.Write and passes in a field that is an array of strings. The code looks similar to:
```
public static void LogFoo(string[] foo) {
    EventSource logger = new EventSource(...);
    logger.Write("...", options, new { Foo = foo });
}
```
This code works without issue when run targeting any version of .net Framework I've tested, and when targeting netcoreapp2.0. It does not work with netcoreapp2.1 with runtime version v2.1.300, instead throwing a System.NotSupportedException: 'Arrays of null-terminated string are not supported.' . While the message is pretty clear, I'm wondering if this functionality regression was intentional?

**Comment by davidmatson** [comment]
Yeah, having a minor version include breaking changes isn't good semver, so I'm hoping that's not intentional.

@brianrob, would you happen to know what the root cause might be? (Since you were the last one to edit [that line](https://github.com/dotnet/coreclr/blame/85374ceaed177f71472cc4c23c69daf7402e5048/src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/TraceLogging/FieldMetadata.cs#L141), though admittedly just changing how the error lookup happens.)

**Comment by vancem** [comment]
including @jorive @noahfalk

**Comment by tommcdon** [comment]
@noahfalk any objections moving this to future?

**Comment by noahfalk** [comment]
I'd still like to get this addressed in 3.0 if we can

**Comment by josalem** [comment]
After some research, changing this in a fully supportable manner (not just a workaround) would involve non-trivial changes to the serialization format used.  This could introduce instability in both the source and reader that is unwanted going into the 3.0 release.  I'm going to move this to future for now.  @brianrob, you're most likely much more familiar with these code paths, if you think this would be a relatively straightforward update to the format, please let me know.

**Comment by noahfalk** [comment]
@josalem - can you elaborate on your investigation results such as what options you looked at and what changes/risks those options cause? Since we've only marked this Future rather than closing it suggests that someone is going to be re-opening this investigation and it would be a shame if they had to start over from the beginning.

**Comment by brianrob** [comment]
@josalem, I suspect you are right that this is not going to be trivial.  It stems from the fact that I was trying to unify the way we serialize strings between TraceLogging ETW and EventPipe so that I could add support for TraceLogging in EventPipe without having to support two deserialization schemes within TraceEvent. TraceLogging ETW previously used counted strings and had support for arrays of counted strings but when I switched things over to NULL-terminated strings, I did not realize, nor have a test case that covered arrays of strings.  The code had always been written to disallow this, but the path was never hit.

I suspect that it is possible to implement this without too much trouble, but I don't know what the encoding would look like.  I don't believe that this would require fundamental changes to EventSource, but would require that we understand what the encoding should look like for supporting this on the ETW side, and then making sure that this works on the EventPipe side as well.

If you want to dig into this more, let me know and I can connect you with the right person who can tell us what the encoding should look like.

**Comment by josalem** [comment]
@noahfalk, as @brianrob points out, it looks to me like the majority of this work is going to be nailing down a stable encoding for arrays of "primitive" types (primitive in regards to what is already serializable by ETW/EventPipe).  That, in and of itself, shouldn't be too difficult (famous last words and all that), but will require updates to the serializers for ETW _and_ EventPipe, as well as updating the deserializers.  

I feel there is higher priority work that will introduce less instability that can be completed with the timeframe we're looking at.  I do think this work _is_ worth implementing, as having arrays of "things" inside events would be very useful.  I'm open to moving it back into 3.0, but I think IPC work should take priority since the current encoding seems to have only recently stabilized.  @brianrob, since there is going to be work surrounding serialization in general I think it would be worth syncing offline to go over the moving parts.

**Comment by brianrob** [comment]
Sure, @josalem.  Happy to chat.
